packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -88,11 +88,145 @@ class AnnotationResult<T> {
 /// [SceneBuilder.build] to obtain a [Scene]. A [Scene] can then be painted
 /// using [dart:ui.FlutterView.render].
 ///
+/// ## Memory
+///
+/// Layers retain resources between frames to speed up rendering. A layer will
+/// retain these resources until all [LayerHandle]s referring to the layer have
+/// nulled out their references.
+///
+/// Layers must not be used after disposal. If a RenderObject needs to maintain
+/// a layer for later usage, it must create a handle to that layer. This is
+/// handled automatically for the [RenderObject.layer] property, but additional
+/// layers must user their own [LayerHandle].
+///
+/// {@tool snippet}
+///
+/// This [RenderObject] is a repaint boundary that pushes an additional
+/// [ClipRectLayer].
+///
+/// ```dart
+/// class ClippingRenderObject extends RenderBox {
+///   final LayerHandle<ClipRectLayer?> _clipRectLayer = LayerHandle<ClipRectLayer?>();
+///
+///   @override
+///   bool get isRepaintBoundary => true; // The [layer] property will be used.
+///
+///   @override
+///   void paint(PaintingContext context, Offset offset) {
+///     _clipRectLayer.layer = context.pushClipRect(
+///       needsCompositing,
+///       offset,
+///       Offset.zero & size,
+///       super.paint,
+///       clipBehavior: Clip.hardEdge,
+///       oldLayer: _clipRectLayer.layer,
+///     );
+///   }
+///
+///   @override
+///   void dispose() {
+///     _clipRectLayer.layer = null;
+///     super.dispose();
+///   }
+/// }
+/// ```
+/// {@end-tool}
 /// See also:
 ///
 ///  * [RenderView.compositeFrame], which implements this recomposition protocol
 ///    for painting [RenderObject] trees on the display.
 abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
+  /// If asserts are enabled, returns whether [dispose] has
+  /// been called since the last time any retained resources were created.
+  ///
+  /// Throws an exception if asserts are disabled.
+  bool get debugDisposed {
+    late bool disposed;
+    assert(() {
+      disposed = _debugDisposed;
+      return true;
+    }());
+    return disposed;
+  }
+  // TODO(dnfield): https://github.com/flutter/flutter/issues/85066
+  final bool _debugDisposed = false;
+
+  /// Set when this layer is appended to a [ContainerLayer], and
+  /// unset when it is removed.
+  ///
+  /// This cannot be set from [attach] or [detach] which is called when an
+  /// entire subtree is attached to or detached from an owner. Layers may be
+  /// appended to or removed from a [ContainerLayer] regardless of whether they
+  /// are attached or detached, and detaching a layer from an owner does not
+  /// imply that it has been removed from its parent.
+  final LayerHandle<Layer> _parentHandle = LayerHandle<Layer>();
+
+  /// Incremenetd by [LayerHandle].
+  int _refCount = 0;
+
+  /// Called by [LayerHandle].
```

Comment:
nit: I'm not sure if we have a guideline for this, but I prefer that docs are standalone (i.e. they talk about what the method does rather than what its used for). Usage docs can get stale easily because nothing prompts the person changing the usage to come here an update the doc. This one may be a special case and `unref` is specifically designed for `LayerHandle` to call.

Reply:
This is really just a pointer for framework developers on a private method. It's not user-facing, and I just put it there so you don't get confused about why a class has a privat emethod it never uses :)